### ISSUE TITLE
Type error with 'extends' constraint and inheritance

### ISSUE THREAD

**Issue by Kais-DkM** [Microsoft/TypeScript#17316]
**TypeScript Version:**  2.4.1

**Code**

```ts
interface A {
  a: number;
}

interface B extends A {
  b: number;
}

function x(b: B): void {}
const y: <C extends A>(c: C) => void = x;
```

**Expected behavior:**
No Type Error

**Actual behavior:**
Type error with message : 
```
Type '(b: B) => void' is not assignable to type '<C extends A>(c: C) => void'.
  Types of parameters 'b' and 'c' are incompatible.
    Type 'C' is not assignable to type 'B'.
      Type 'A' is not assignable to type 'B'.
        Property 'b' is missing in type 'A'.
```

**Comment**
Weirdly, it doesn't show any errors on Playground, but it shows errors when compiled on local project.

**Comment by jcalz** [comment]
This is intended behavior.

Playground is currently using TypeScript 2.3.3.  [Stricter checking for generic functions](https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#stricter-checking-for-generic-functions) was introduced in TypeScript 2.4.

The error message is a good one.   If I change your implementation of `x()` to actually use the `b` parameter:

```ts
function x(b: B): void { 
  console.log(b.b.toString()); // expects b.b to be defined
} 
const y: <C extends A>(c: C) => void = x;  
```
Then you can see why assigning `x` to `y` is flagged.  You can apparently call `y()` on any `A`, but this will blow up at runtime unless the parameter is a `B`:
```ts
y({'a': 2});  // no error in TS, but TypeError at runtime
```

**Comment by Kais-DkM** [comment]
@jcalz 
I'm still confused about your EDIT part, but I think my problem is answered.
Thank you for your help.
